### PR TITLE
Fixes AI integrity restorer dropping AI core on decon

### PR DESCRIPTION
**Pull request by TetraK1**
## About The Pull Request

AI mob would get dropped when deconning or destroying the integrity restorer.  The previous check to prevent screwdriver deconstruction with an ai inside was using the wrong proc, and destroying it by force wasn't accounted for.  Now the screwdriver check works, and it can't be deconned with an ai inside.  If the console is destroyed into a computer frame, AI is removed permanently.

## Why It's Good For The Game

Previously you could stick an intellicard into the console, then decon or destroy it, and get a free AI core.  I was also promised nedbux.

## Changelog
:cl:Tetr4
fix: AI integrity restorer no longer drops a free AI core on deconstruction/destruction.
/:cl:

### CODE REVIEW

**Comment by zxaber** [comment]
For the record, this doesn't affect the console that takes the whole card, right?

**Comment by TetraK1** [comment]
No, afaik that's just a modular console with a card slot component, which just drops the slot component with intellicard still inside when destroyed.

**Comment by zxaber** [comment]
Well that's good at least.

Would it be worth pursuing an alternative idea where the AI stays in the circuit board? In that scenario, the AI would be dead as soon as the console is smashed, but could at least be revived if the board isn't lost.

I guess having the repair console hold some danger for the AI isn't the end of the world, but it _would_ suck if a freak incident ended the round for you. Imagine being straight-up deleted by a immovable rod (especially since the AI can't be removed mid-restore).

**Comment by ExcessiveUseOfCobblestone** [comment]
so if it still drops on destroy does this actually fix the bug?

**Comment by zxaber** [comment]
There's two separate AI restore consoles. One is a standalone, and the other is part of the RD's modular computer. Using an AI card on the modular computer will take the whole card in, so the AI exists on the card the entire time. The standalone does not take the card, and just has the AI.

Far as TetraK1 says, this only affects the standalone, which does currently drop the AI mob on the floor (and it looks like a new core because the AI core is actually just the mob).

**Comment by ExcessiveUseOfCobblestone** [comment]
I didn't see the on_destroy mybad.

File: code/game/machinery/computer/aifixer.dm [hunk 1]
```diff
@@ -148,3 +148,7 @@
 			to_chat(user, "<span class='boldannounce'>ERROR</span>: Reconstruction in progress.")
 		else if (!occupier)
 			to_chat(user, "<span class='boldannounce'>ERROR</span>: Unable to locate artificial intelligence.")
+
+obj/machinery/computer/aifixer/on_deconstruction()
+	if(occupier)
+		occupier.Destroy()
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/machinery/computer/aifixer.dm]: you might want to null out occupier after this?

Comment (SpaceManiac) [reply code/game/machinery/computer/aifixer.dm]: `Destroy` is the wrong call to use here. In general it should be `qdel(occupier)` to delete something, and in this case to set it to null afterwards, `QDEL_NULL(occupier)`

File: code/game/machinery/computer/aifixer.dm [hunk 2]
```diff
@@ -148,3 +148,7 @@
 			to_chat(user, "<span class='boldannounce'>ERROR</span>: Reconstruction in progress.")
 		else if (!occupier)
 			to_chat(user, "<span class='boldannounce'>ERROR</span>: Unable to locate artificial intelligence.")
+
+obj/machinery/computer/aifixer/on_deconstruction()
```
Comment (SpaceManiac) [inline code/game/machinery/computer/aifixer.dm]: Our style is to start lines like these with a `/`:
```suggestion
/obj/machinery/computer/aifixer/on_deconstruction()
```

**Review by SpaceManiac — CHANGES_REQUESTED** [review]
Thanks for the PR and welcome to contributing! Just a couple of small notes, feel free to ask if something is unclear.

**Comment by TetraK1** [comment]
I'll do the circuit-board-holding-AI thing another PR, won't be able to for a few days.

**Pull request merged**